packages/upgrade/README.md

```diff
@@ -0,0 +1,107 @@
+
+So, when inspecting the upgrade pipeline, the seperation of concerns is clear:
+
+- The entire upgrade process and any on-chain actions are execute via `forge script` and the logic is written in Solidity
+- The typescript wrapper is only responsible for supplying the correct data to the `forge script` and storing artifacts for posterity
+
+### Next Steps
+
+As `forge script` matures, the typescript wrapper will get thinner, as more functionality is moved into Solidity (e.g multi-chain deployments). The current biggest limitations are an easy way to input and output data from `forge script`so that the upgrade pipeline can:
+
+- Read a config file
+- Execute upgrade based on config file
+- Store artifacts that are part of the output of the upgrade
+
+Forge script is under heavy development, so I expect the typescript wrapper to be considerably thinner in the next Nomad Upgrade.
+
+## Installation
+
+- Install [Foundry](https://github.com/foundry-rs/foundry) and update it to the latest version via `foundryup`
+- Install [ts-node](https://www.npmjs.com/package/ts-node)
+
+## Usage
+
+```
+ts-node ./scripts/upgrade.ts -c <path_to_config> <functionality> -d domain1 -d domain2
+
+HELP
+--config: path to config file
+--domain: Domains to which the functionality will be applied, expects a domain name (e.g ethereum, evmos, etc.)
+--all: Do all domains
+--resume: In case upgrade has run before but it wasn't complete for some reason (e.g RPC failing), you can resume the transactions. Please make sure that the account nonce hasn't been incremented in the meantime (e.g by sending a tx)
+
+COMMANDS
+upgrade: Upgrade the protocol
+executeCallbatch: Execute a callBatch on some domain
+printGovActions: Print the bytecode calldata in hex format for executing `executeGovernanceActions()`, encoded with the correct args
+```
+
+**Note:**
+
+Due to `oclif` bad flag parsing, when using the flag `-d`, all space-seperated strings after the flag will be parsed as flag arguments.
+
+```bash
+scripts/upgrade.ts -d ethereum upgrade ==> domains = ['ethereum', 'upgrade']
+```
+
+For that reason, we need to use the `-d` flag at the end of the command:
+
+```bash
+scripts/upgrade.ts printGovActions -d ethereum
+```
+
+### Output & Artifacts
+
+- Artifacts & raw output are placed in `upgrade-artifacts`, a directory that is created at the root of `upgrade`
+- Artifacts are extracted from the raw output. The raw output contains ALL the relevant information of a Nomad Upgrade (Nomgrade)
+- All the files are placed under their respective domain (e.g `upgrade-artifacts/ethereum`)
+- Raw output is stored for every command that is executed
+- Artifacts are placed in `artifacts.json` - `executeCallBatchCall`: CallData to `executeCallBatch()` on the Governance Router of a domain, after it has received the Governance Message. Abi encoded function signature and arguments
```

Comment:
all of the functionality around call batches is basically duplicated with sdk-govern. we want to maintain **one** code path for doing this.